### PR TITLE
controller/engine: add queueing index method to NamedController

### PR DESCRIPTION
**Pull request by sttts**
### Description of your changes

Using `GetCache()`'s `IndexField` method will use the passed context for the life-cycle of informers started at that moment. This is usually not what is desired, but the life-cycle of the cache should match the life-cycle of the controller, and this applies to all informers of course.

This PR adds a `IndexField` method to queue up index creation until the controller has been started.

Note that the `GetCache()` method is still useful to e.g. wire it up into other components that want to read from the cache later.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Tested in Crossplane fixing realtime composition issues.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: pkg/controller/engine.go [hunk 1]
```diff
@@ -187,15 +187,33 @@ func (e *Engine) Start(name string, o controller.Options, w ...Watch) error {
...
+	// GetCache returns the cache used by the named controller. If the
+	// controller hasn't been started, neither has the cache.
+	//
+	// Warning: Be careful when calling methods with a context before the
+	// controller has been started. The pass context will be used to e.g. start
+	// informers, and the life-cycle will differ from that of the controller. In
+	// particular, use IndexField instead to add indexes to the cache before the
+	// controller is started.
 	GetCache() cache.Cache
+	IndexField(obj client.Object, field string, extractValue client.IndexerFunc)
```
Comment (negz) [inline pkg/controller/engine.go:201]: Could you add a doc string for this method, since the others all have them.

File: pkg/controller/engine.go [hunk 2]
```diff
@@ -281,7 +305,18 @@ func (c *namedController) Start(ctx context.Context) error {
 	return nil
 }
 
+// IndexField queues an index for addition to the cache on start.
```
Comment (negz) [inline pkg/controller/engine.go:308]: Would `QueueIndexField` or `IndexFieldOnStart` or something similar make this more self-descriptive?

What happens if this is called after the controller/cache is started? Nothing I guess? (Worth mentioning in doc string.)

File: pkg/controller/engine.go [hunk 3]
```diff
@@ -187,15 +187,33 @@ func (e *Engine) Start(name string, o controller.Options, w ...Watch) error {
...
 type namedController struct {
 	name string
 	e    *Engine
 	ca   cache.Cache
 	ctrl controller.Controller
+
+	indexes []index
+}
+
+type index struct {
```
Comment (negz) [inline pkg/controller/engine.go:213]: Would it make sense to export this type and have Start take a variadic array of it? I'm not sure how this is called, so I'm happy if it's more ergonomic to use the method approach instead.

e.g.:

```go
c.Start(ctx,
    Index{Object: o, Field: "spec.hi", ExtractValue: ifn},
    Index{Object: o, Field: "spec.hello", ExtractValue: ifn2})
```

File: pkg/controller/engine.go (hunk 2 above)
Comment (sttts) [reply pkg/controller/engine.go:308]: Good points.

Comment (sttts) [reply pkg/controller/engine.go:308]: Actually, standby with this PR. The context passed to `GetInformer` and with that to `IndexField` is only used for waiting, not for the informer life-cycle. I think this was a misunderstanding.

**Comment by negz** [comment]
https://github.com/crossplane/crossplane-runtime/pull/689 moved the engine implementation to c/c. The implementation is being reworked there so this PR may no longer be relevant but either way it won't happen in this repo. 🙂

**Pull request closed without merging**